### PR TITLE
feat: introduce new columns for session ratings & switch to ember table

### LINKED ISSUE #3195: Add required columns to 'Manage Events/Sessions' for rating
Sub-issue of https://github.com/fossasia/open-event-frontend/issues/2959

We give any role that has the right to rate
   - the option to choose any star from 1 to 5

Also then we can have two more column showing
   - average rating
   - number of people who rated

Columns Required - `| Rate session | Avg Rating | No. of people rated |`

**Additional Context**
On it

### PR DESCRIPTION
**Pull request by shreyanshdwivedi**
Fixes #3195 

#### Short description of what this resolves:
Adds columns - `| Rating | Avg Rating | No. of ratings |`

#### Changes proposed in this pull request:
 - Modified `updateRating` function
 - Introduced 2 new columns - `Avg Rating` and `No. of ratings`

Screenshot_2019-08-01 Confirmed Sessions AddOwnerEvent Events Open Event

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: app/controllers/events/view/sessions/list.js [hunk 1]
```diff
@@ -1,6 +1,19 @@
 import Controller from '@ember/controller';
+import { computed } from '@ember/object';
 
 export default Controller.extend({
+  ratedSessions: computed('model.data.@each.feedbacks', async function() {
+    let ratedSessions = [];
+    await this.model.data.forEach(async session => {
+      await session.feedbacks.forEach(async feedback => {
+        let { user } = await feedback;
```
Comment (mrsaicharan1) [inline app/controllers/events/view/sessions/list.js]: Write a separate computed property which would return feedbacks and pass it in as a watchable property. In that way, you wouldn't have to use async//await again while looping through feedbacks

File: app/controllers/events/view/sessions/list.js [hunk 2]
```diff
@@ -1,6 +1,15 @@
...
+import { computed } from '@ember/object';
 
 export default Controller.extend({
+  ratedSessions: computed('model.feedbacks.@each.session', function() {
+    let ratedSessions = [];
+    this.model.feedbacks.forEach(feedback => {
+      let { session } = feedback;
+      ratedSessions.push(session);
+    });
+    return ratedSessions;
```
Comment (niranjan94) [inline app/controllers/events/view/sessions/list.js]: This is can simplified with [`mapBy`](http://api.emberjs.com/ember/3.10/functions/@ember%2Fobject%2Fcomputed/mapBy) ...

```js
ratedSessions: mapBy('model.feedbacks', 'session')
```

**Comment by kushthedude** [comment]
@shreyanshdwivedi Status?

**Comment by shreyanshdwivedi** [comment]
Been busy with more pressing bugs. Will work on it once they are resolved.

File: app/controllers/events/view/sessions/list.js [hunk 3]
```diff
@@ -169,31 +187,31 @@ export default Controller.extend({
...
-            comment,
-            session,
-            user
-          });
-          await feedback.save();
-          this.notify.success(this.l10n.t('Session feedback has been created successfully.'));
-        }
+        this.set('isLoading', true);
+        const user = this.authManager.currentUser;
+        const comment = '';
```
Comment (niranjan94) [inline app/controllers/events/view/sessions/list.js]: Comment and user need not be stored into variables here. Can be directly used.

File: app/routes/events/view/sessions/list.js [hunk 4]
```diff
@@ -60,12 +60,33 @@ export default Route.extend({
       'page[size]' : 10
     };
 
-    let store = this.modelFor('events.view');
 
+    let store = this.modelFor('events.view');
     let data =  await store.query('sessions', queryObject);
 
+    const { currentUser } = this.authManager;
+    let feedbacks = await currentUser.query('feedbacks', {
```
Comment (niranjan94) [inline app/routes/events/view/sessions/list.js]: You can use `this.authManager.currentUser` directly here

File: app/controllers/events/view/sessions/list.js (hunk 2 above)
Comment (shreyanshdwivedi) [reply app/controllers/events/view/sessions/list.js]: @niranjan94 I resumed my work on this PR. I have one query that ratedSessions is of `ArrayProxy` type and so cannot be directly used in template. Also I cannot use await here directly. Can you please point out how can this proxy be resolved?

Comment (niranjan94) [reply app/controllers/events/view/sessions/list.js]: array proxy can be directly used as an array ... that's the whole point of an array proxy.

**Comment by shreyanshdwivedi** [comment]
I've added the final gif to the PR description. Please review
@niranjan94 @CosmicCoder96 @uds5501 @kushthedude @mrsaicharan1

File: app/controllers/events/view/sessions/list.js [hunk 5]
```diff
@@ -1,6 +1,9 @@
 import Controller from '@ember/controller';
+import { mapBy } from '@ember/object/computed';
 
 export default Controller.extend({
+  ratedSessions: mapBy('model.feedbacks', 'session.id'),
+
   columns: [
```
Comment (niranjan94) [inline app/controllers/events/view/sessions/list.js]: Switch the table to use ember-tables

File: app/controllers/events/view/sessions/list.js [hunk 6]
```diff
@@ -1,6 +1,9 @@
 import Controller from '@ember/controller';
+import { mapBy } from '@ember/object/computed';
 
 export default Controller.extend({
```
Comment (niranjan94) [inline app/controllers/events/view/sessions/list.js]: Switch to es6 classes

File: app/controllers/events/view/sessions/list.js [hunk 7]
```diff
@@ -1,218 +1,243 @@
 import Controller from '@ember/controller';
+import { action } from '@ember/object';
 import { mapBy } from '@ember/object/computed';
+import EmberTableControllerMixin from 'open-event-frontend/mixins/ember-table-controller';
 
-export default Controller.extend({
-  ratedSessions: mapBy('model.feedbacks', 'session.id'),
+export default class extends Controller.extend(EmberTableControllerMixin) {
+  ratedSessions= mapBy('model.feedbacks', 'session.id');
```
Comment (niranjan94) [inline app/controllers/events/view/sessions/list.js]: Use decorator

File: app/controllers/events/view/sessions/list.js [hunk 8]
```diff
@@ -1,218 +1,243 @@
...
 
-  columns: [
+  columns= [
```
Comment (niranjan94) [inline app/controllers/events/view/sessions/list.js]: Should be a computed getter. Only then will action bindings work properly

File: app/controllers/events/view/sessions/list.js [hunk 9]
```diff
@@ -1,218 +1,243 @@
...
+    } catch (error) {
+      this.notify.error(this.l10n.t(error.message));
+    }
+    this.set('isLoading', false);
+  }
+
+  @action
+  async acceptProposal(session, sendEmail) {
+    try {
       session.set('sendEmail', sendEmail);
```
Comment (niranjan94) [inline app/controllers/events/view/sessions/list.js]: use setProperties

File: app/controllers/events/view/sessions/list.js [hunk 10]
```diff
@@ -1,218 +1,243 @@
...
+    }
+    this.send('refreshRoute');
+    this.set('isLoading', false);
+  }
+
+  @action
+  async confirmProposal(session, sendEmail) {
+    try {
       session.set('sendEmail', sendEmail);
       session.set('state', 'confirmed');
```
Comment (niranjan94) [inline app/controllers/events/view/sessions/list.js]: setProperties

File: app/controllers/events/view/sessions/list.js (hunk 7 above)
Comment (shreyanshdwivedi) [reply app/controllers/events/view/sessions/list.js]: Do we have to use this like other computed properties?
```js
@computed()
get ratedSessions() {
    return mapBy('model.feedbacks', 'session.id');
}
```

Comment (niranjan94) [reply app/controllers/events/view/sessions/list.js]: No ...

```js
@mapBy('model.feedbacks', 'session.id')
ratedSessions
```

@shreyanshdwivedi read the docs please -> http://api.emberjs.com/ember/3.10/functions/@ember%2Fobject%2Fcomputed/mapBy

**Comment by shreyanshdwivedi** [comment]
There is a bug in this implementation. When I rate a session the refresh isn't working properly. It is recorded to the server but the rows aren't loaded and so when I rate it again, the FE is sending to create a session rather than updating it so new instance is created on server. @niranjan94 what could be the problem here?

File: app/controllers/events/view/sessions/list.js [hunk 11]
```diff
@@ -1,204 +1,269 @@
...
+        isSortable      : true,
+        headerComponent : 'tables/headers/sort'
+      },
+      {
+        name      : 'Track',
+        valuePath : 'track.name'
+      },
+      {
+        name      : 'Type',
+        valuePath : 'sessionType.name'
```
Comment (shreyanshdwivedi) [inline app/controllers/events/view/sessions/list.js]: @CosmicCoder96 I tried to display `sessionType.name` and `track.name` using this but at first load it is missing. As soon as I trigger any action such as change tab or rate session or anything else, both of them appears.
Should I add a new cell template file and pass a computed property to it and display this?

Comment (shreyanshdwivedi) [reply app/controllers/events/view/sessions/list.js]: This action triggering is taking time and other click for other actions too, e.g. changing tab

**Comment by shreyanshdwivedi** [comment]
@CosmicCoder96 there are few issues I'm facing -
 - The refreshRoute is not loading the FE instances correctly. Like when I lock/unlock a session, the icon doesn't changes (but the data is sent to and recorded by server properly)
 - Similarly, when I rate a session, a new feedback is created but the FE doesn't reload properly and so when I update rating, it is still being created (rather than being updated).
 - When I use `this.asArray(store.query('sessions', queryString)`, I'm getting error - `Unhandled promise rejection Error: "Assertion Failed: value must have an array of children"`. That's why I'm using `let data = (await store.query('sessions', queryString)).toArray();`

Please help me out with these issues.

**Comment by CosmicCoder96** [comment]
I assume this will be done in ember tables PR of sessions table now.

**Comment by shreyanshdwivedi** [comment]
@CosmicCoder96 Actually this PR was for introduction of new columns and actions related to rating a session. I've already completed major part of it.
If you want you can move forward with refactoring it, I'll just add columns and actions for now?

**Comment by CosmicCoder96** [comment]
@shreyanshdwivedi I got confused, you have changed the table here itself so just proceed here.

**Comment by shreyanshdwivedi** [comment]
I need one more help @CosmicCoder96 https://github.com/fossasia/open-event-frontend/pull/3198#issuecomment-515784574

Actually when we performed any action in ember-model-table then after rendering refreshRoute the icons also changed. But now the icons don't change, not here not in any other table.

**Comment by shreyanshdwivedi** [comment]
This PR is ready for review. @CosmicCoder96 @kushthedude @uds5501 @mrsaicharan1 

There is a general drawback of the ember table which I felt, that after triggering an action, the FE isn't updated. It happens everywhere we used ember-table. So, when feedback is created, the FE doesn't change and the same patch is rendered which is associated with `createRating` action. And so, if we don't manually refresh the page after creating feedback, duplicate feedbacks are created.

**Comment by kushthedude** [comment]
> This PR is ready for review. @CosmicCoder96 @kushthedude @uds5501 @mrsaicharan1
> 
> There is a general drawback of the ember table which I felt, that after triggering an action, the FE isn't updated. It happens everywhere we used ember-table. So, when feedback is created, the FE doesn't change and the same patch is rendered which is associated with `createRating` action. And so, if we don't manually refresh the page after creating feedback, duplicate feedbacks are created.

@shreyanshdwivedi Try triggering ‘RefreshRoute’  action after the ratings input is taken.

**Comment by shreyanshdwivedi** [comment]
@kushthedude in today's meeting, @CosmicCoder96 told that this issue is persistent in almost every table we are using for now. He'll handle this in another PR
@CosmicCoder96 please review this PR

**Comment by kushthedude** [comment]
@shreyanshdwivedi Are the missing columns expected?
image

**Comment by shreyanshdwivedi** [comment]
@kushthedude they are appearing for me. Can you try by clicking on `All` tab again? Maybe they are not loaded properly for you.

**Comment by kushthedude** [comment]
> @kushthedude they are appearing for me. Can you try by clicking on `All` tab again? Maybe they are not loaded properly for you.

@shreyanshdwivedi I did it, Its strange when I refresh the page `Track and Type` is not loaded but when I click all again they loads. Please test it

**Comment by shreyanshdwivedi** [comment]
@kushthedude I tested this and encountered this earlier but to fix it I introduced new cell-name.hbs template
But I think it's a problem with ember table then because the data which is shown is not a computed data.
@CosmicCoder96 can you please review this PR once

**Comment by kushthedude** [comment]
@shreyanshdwivedi One more thing, When I refresh the page at `Sessions Table`, I see the `Manage Events Nav-Bar` again, Something might be wrong at `Routes`, Please see the reproducibility.
image

File: app/templates/components/ui-table/cell/cell-name.hbs [hunk 12]
```diff
@@ -0,0 +1,3 @@
+{{#if record}}
+  {{record.name}}
```
Comment (CosmicCoder96) [inline app/templates/components/ui-table/cell/cell-name.hbs]: oversight? This does not make sense.

Comment (shreyanshdwivedi) [reply app/templates/components/ui-table/cell/cell-name.hbs]: Actually I added this template just to fix issue mentioned here - https://github.com/fossasia/open-event-frontend/pull/3198#issuecomment-518772521

But it isn't helping. I used `sesstionType.name` and `track.name` earlier in column in list.js

Comment (kushthedude) [reply app/templates/components/ui-table/cell/cell-name.hbs]: @shreyanshdwivedi Try using extraValuePaths as done in other tables to use other properties of a variable.
https://github.com/fossasia/open-event-frontend/blob/ac210a331d034d9ce1e98380923271656c334f53/app/controllers/admin/messages/list.js#L30

Comment (shreyanshdwivedi) [reply app/templates/components/ui-table/cell/cell-name.hbs]: No help @kushthedude 
@CosmicCoder96 reverting it to previous state i.e. passing `valuePath     : 'track.name'` and `valuePath     : 'sessionType.name'` in respective columns.

**Comment by shreyanshdwivedi** [comment]
@CosmicCoder96 please have a look now. There are a few glitches but it exist in all the tables for now. Can we move ahead with this PR as it blocks one of the enhancement?

**Comment by shreyanshdwivedi** [comment]
@uds5501 @mrsaicharan1 @kushthedude please review this. We need to move forward with this PR for further enhancement so review this on priority please.
Thank You

**Pull request merged**